### PR TITLE
Add planar translation and colorize selected constraint ;

### LINKED ISSUE #46: Torus shape for the rotation gizmos.
Prettier shape.

### LINKED ISSUE #47: Add scale gizmo.
weeee. Scale !

### PR DESCRIPTION
**Pull request by hoshiryu**
* **Please check if the PR fulfills these requirements**
- [x] Is the Pull-Request done against the right branch:
  - `master`: for hotfixes not impacting the API
  - `master-v1`: for changes related to the Radium Stable Release V1.
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

Be aware that the PR request cannot be accepted if it doesn't pass the Continuous Integration tests.

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Fix #46 
Fix #47
 - Add plane manipulators for the translation gizmo, thus allowing in-plane translations.
 - Add scaling gizmo.
 - Fix rotation gizmo when rotation plane is orthogonal to image plane.
 - Rescale all gizmos RO.
 - Colors the selected constraint in yellow.
 - Improve keymapping management.
 - Also fix skeleton bone transformations from the animation component.

* **What is the current behavior?** (You can also link to an open issue here)
 - when the rotation plane is orthogonal to the image plane, the rotation gizmo doesn't work.
 - Regarding skeleton bone transformations, the parent is rotated to follow its child.
However, in the case of multiple children, this shouldn't be the case.

* **What is the new behavior (if this is a feature change)?**
 - In order to perform in-plane translation, click & drag the square of the translation gizmo.
 - Same for scale.
 - when the rotation plane is orthogonal to the image plane, the rotation gizmo now works.
 - Regarding skeleton bone transformations, the parent is rotated only if it has only one child.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

**Comment by nmellado** [comment]
Can you please explain how to use this feature ? Is it bind in the main app ?

File: src/GuiBase/Viewer/Gizmo/ScaleGizmo.hpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+#include <GuiBase/Viewer/Gizmo/Gizmo.hpp>
+namespace Ra {
+namespace Gui {
+class ScaleGizmo : public Gizmo {
+  public:
+    // Draw a frame with : 3 RGB arrows
+    ScaleGizmo( Engine::Component* c, const Core::Transform& worldTo, const Core::Transform& t,
+                Mode mode );
+
+    virtual void updateTransform( Gizmo::Mode mode, const Core::Transform& worldTo,
```
Comment (nmellado) [inline src/GuiBase/Viewer/Gizmo/ScaleGizmo.hpp]:
```suggestion
    void updateTransform( Gizmo::Mode mode, const Core::Transform& worldTo,
```

File: src/GuiBase/Viewer/Gizmo/ScaleGizmo.hpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+                                  const Core::Transform& t ) override;
+    virtual void selectConstraint( int drawableIndex ) override;
```
Comment (nmellado) [inline src/GuiBase/Viewer/Gizmo/ScaleGizmo.hpp]:
```suggestion
    void selectConstraint( int drawableIndex ) override;
```

File: src/GuiBase/Viewer/Gizmo/ScaleGizmo.hpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    virtual void setInitialState( const Engine::Camera& cam,
```
Comment (nmellado) [inline src/GuiBase/Viewer/Gizmo/ScaleGizmo.hpp]:
```suggestion
    void setInitialState( const Engine::Camera& cam,
```

File: src/GuiBase/Viewer/Gizmo/ScaleGizmo.hpp [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+                                  const Core::Vector2& initialXY ) override;
+    virtual Core::Transform mouseMove( const Engine::Camera& cam, const Core::Vector2& nextXY,
```
Comment (nmellado) [inline src/GuiBase/Viewer/Gizmo/ScaleGizmo.hpp]:
```suggestion
    Core::Transform mouseMove( const Engine::Camera& cam, const Core::Vector2& nextXY,
```

File: src/GuiBase/Viewer/Gizmo/RotateGizmo.hpp [hunk 5]
```diff
@@ -5,28 +5,31 @@
...
 
   private:
-    Core::Vector2 m_initialPix;
-    int m_selectedAxis;
-    bool m_start;
-    bool m_stepped;
-    float m_totalAngle;
-    Core::Matrix3 m_initialRot;
+    Core::Vector2 m_initialPix; ///< The pixel location when edition starts.
+    int m_selectedAxis;         ///< The axis to ratate around.
```
Comment (nmellado) [inline src/GuiBase/Viewer/Gizmo/RotateGizmo.hpp]:
```suggestion
    int m_selectedAxis;         ///< The axis to rotate around.
```

**Comment by hoshiryu** [comment]
There is a little bug when using local gizmos, will be handled soon!

**Comment by hoshiryu** [comment]
Ready

**Comment by hoshiryu** [comment]
Ready for review

File: src/GuiBase/Utils/KeyMappingManager.cpp [hunk 6]
```diff
@@ -4,6 +4,15 @@
 
 namespace Ra {
 namespace Gui {
+
+//char const* const KeyMappingManager::KeyMappingActionNames[] = {
```
Comment (MathiasPaulin) [inline src/GuiBase/Utils/KeyMappingManager.cpp]:
```suggestion
```

File: src/GuiBase/Utils/KeyMappingManager.cpp [hunk 7]
```diff
@@ -4,6 +4,15 @@
...
+const std::string KeyMappingManager::KeyMappingActionNames[] = {
```
Comment (MathiasPaulin) [inline src/GuiBase/Utils/KeyMappingManager.cpp]: This is redundant with the services offered by ``Q_ENUM( KeyMappingAction )`` (line 56 of KeyMappingManager.hpp) and the operation ``m_metaEnumAction.valueToKey(action)`` that convert an enum value of type ``KeyMappingAction ``  to a ``const char *`` (see Qt documentation). The ``QMetaEnum m_metaEnumAction`` is required for parsing the xml configuration file. Nevertheless, we can keep this as its usage is Qt independent and do not need to have access to the required ``QMetaEnum`` and  the``friend std::ostream & operator <<(std::ostream& out, KeyMappingManager::KeyMappingAction action)`` operator could be used everywhere.

File: src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp [hunk 8]
```diff
@@ -19,13 +19,19 @@
 
 namespace Ra {
 namespace Gui {
+
+inline void colorMesh( std::shared_ptr<Engine::Mesh> mesh, const Core::Color& color ) {
```
Comment (MathiasPaulin) [inline src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp]: Why is this function local to ``RotateGizmo.cpp``. The same is define also in ScaleGizmo.cpp and ``TranslateGizmo.cpp`` and ...
This method must be define only once, e.g. in ``Engine/Renderer/Mesh/Mesh.hpp`` as a static inline public method.

File: src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp [hunk 9]
```diff
@@ -140,7 +160,9 @@ Core::Transform RotateGizmo::mouseMove( const Engine::Camera& cam, const Core::V
                                             m_worldTo * rotationAxis, hits2 );
 
         Core::Vector2 nextXY_ = nextXY;
-        if ( hit1 && hit2 )
+        Scalar angle;
+        // standard check  +  guard against precision issues
+        if ( hit1 && hit2 && hits1[0] > 0.2 && hits2[0] > 0.2 )
```
Comment (MathiasPaulin) [inline src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp]: Why the constant ``0.2`` ? This is a very big epsilon ....

File: src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+        Core::Vector3 endPoint;
+        if ( findPointOnPlane( cam, origin, scaleDir, m_initialPix + nextXY, endPoint ) )
+        {
+            const Core::Vector3 a = endPoint - m_startPos;
+            if ( a.squaredNorm() < 1e-3 )
+            {
+                return m_transform;
+            }
+            const Core::Vector3 b = m_startPoint - m_startPos;
+            Core::Vector3 scale = a.norm() / b.norm() * Core::Vector3::Ones();
```
Comment (MathiasPaulin) [inline src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp]: This constrains the scale to be positive. At use, this results to strange behavior when one might expect to obtain a symmetric object when the scale become negative and see only a kind of "bouncing" object ...

File: src/GuiBase/Viewer/Gizmo/TranslateGizmo.cpp [hunk 11]
```diff
@@ -19,15 +19,26 @@
 
 namespace Ra {
 namespace Gui {
+
+inline void colorMesh( std::shared_ptr<Engine::Mesh> mesh, const Core::Color& color ) {
```
Comment (MathiasPaulin) [inline src/GuiBase/Viewer/Gizmo/TranslateGizmo.cpp]: See comment on RotateGizmo

File: src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+#include <Engine/Renderer/Material/Material.hpp>
+#include <Engine/Renderer/RenderTechnique/RenderTechnique.hpp>
+#include <Engine/Renderer/RenderTechnique/ShaderConfigFactory.hpp>
+
+#include <GuiBase/Utils/Keyboard.hpp>
+
+namespace Ra {
+namespace Gui {
+
+inline void colorMesh( std::shared_ptr<Engine::Mesh> mesh, const Core::Color& color ) {
```
Comment (MathiasPaulin) [inline src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp]: See comment on RotateGismo

File: src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp (hunk 9 above)
Comment (hoshiryu) [reply src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp]: The thing is that when fitting the camera, some rotation planes are orthogonal to the image plane.
However, the camera doesn't lie exactly on the plane, which makes the rayCast hit the plane very close to the camera when moving the mouse below the plane.
The more you zoom towards the object, the more this value will grown. I found out that at the most zoomed, the value is at 0.6

File: src/GuiBase/Utils/KeyMappingManager.cpp (hunk 7 above)
Comment (hoshiryu) [reply src/GuiBase/Utils/KeyMappingManager.cpp]: What about this?

File: src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp (hunk 9 above)
Comment (hoshiryu) [reply src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp]: Can someone test the rotation gizmo an very small/large scale scenes to check this "add hoc" value  is ok?

File: src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp (hunk 10 above)
Comment (hoshiryu) [reply src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp]: updated doc

File: src/GuiBase/Utils/KeyMappingManager.cpp (hunk 7 above)
Comment (MathiasPaulin) [reply src/GuiBase/Utils/KeyMappingManager.cpp]: We will keep both solutions for now.

File: src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp (hunk 9 above)
Comment (MathiasPaulin) [reply src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp]: Some crash in tangent situations, fixed by commit ff4ed82

File: src/Engine/Renderer/Mesh/Mesh.inl [hunk 13]
```diff
@@ -39,14 +39,21 @@ void Mesh::setDirty( const Mesh::MeshData& type ) {
...
 }
+
 void Mesh::setDirty( const Mesh::Vec4Data& type ) {
     m_dataDirty[MAX_MESH + MAX_VEC3 + type] = true;
     m_isDirty = true;
 }
 
+void Mesh::colorize( const Core::Color& color ) {
+    Core::Vector4Array colors( getGeometry().vertices().size(), color );
+    addData( Engine::Mesh::VERTEX_COLOR, colors );
```
[the remark this answers is not in the document]
Comment (hoshiryu) [reply src/Engine/Renderer/Mesh/Mesh.inl]: I don't see the problem

**Comment by MathiasPaulin** [comment]
@hoshiryu can you resolve the conflict in Viewer.cpp so that we could make the final review ?

**Comment by hoshiryu** [comment]
Which conflict?

**Comment by MathiasPaulin** [comment]
@hoshiryu see the summary in the PR page on github

**Pull request merged**